### PR TITLE
refactor(query): extract number types

### PR DESCRIPTION
**Pull request by andylokandy**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- Extract number types from `DataType`, `Scalar`, `ScalarRef`, `Column`, and `ColumnBuilder` to individual `NumberDataType`, `NumberScalar`, `NumberColumn`, and `NumberColumnBuilder`
- Wrap all `f32` and `f64` into `OrderedFloat<f32>` and `OrderedFloat<f64>`

Related to https://github.com/datafuselabs/databend/issues/6547

### CODE REVIEW

File: src/query/functions-v2/tests/it/scalars/string.rs [hunk 1]
```diff
@@ -455,22 +483,35 @@ fn test_oct(file: &mut impl Write) {
...
         ),
         (
             "d",
-            DataType::Float64,
-            Column::from_data(vec![10f64, -20f64, 30f64]),
+            DataType::Number(NumberDataType::Float64),
+            Column::from_data(
+                vec![10f64, -20f64, 30f64]
+                    .into_iter()
+                    .map(OrderedFloat)
```
Comment (sundy-li) [inline src/query/functions-v2/tests/it/scalars/string.rs:512]: maybe we can provide a impl for `Clumn::from_data(vec![10f64, -20f64, 30f64])`

File: src/query/codegen/Cargo.toml [hunk 2]
```diff
@@ -20,3 +20,5 @@ common-expression = { path = "../expression" }
 
 serde = { version = "1.0.137", features = ["derive"] }
 serde_json = "1.0.81"
+ordered-float = { git = "https://github.com/andylokandy/rust-ordered-float.git", branch = "as", features = ["serde"] }
```
Comment (BohuTANG) [inline src/query/codegen/Cargo.toml:23]: Can we publish a crate or move it to `https://github.com/datafuse-extras`?

Comment (andylokandy) [reply src/query/codegen/Cargo.toml:23]: It's just temporary for https://github.com/reem/rust-ordered-float/pull/110

Comment (andylokandy) [reply src/query/codegen/Cargo.toml:23]: What's the convention for such situations?

Comment (BohuTANG) [reply src/query/codegen/Cargo.toml:23]: OK， we do not recommend to use a personal git repo in databend :)

**Comment by BohuTANG** [comment]
https://github.com/datafuselabs/databend/runs/8282540454?check_suite_focus=true#step:4:1810
```
test scalars::string::test_string ... FAILED

failures:

---- scalars::string::test_string stdout ----

Goldenfile diff for "string.txt":
To regenerate the goldenfile, run
    REGENERATE_GOLDENFILES=1 cargo test
```

**Comment by BohuTANG** [comment]
@mergify update

**Comment by BohuTANG** [comment]
`test_sqllogic_base_cluster_linux` can't passed(I have re-run it many times).

**Review by sundy-li — APPROVED** [review]

**Pull request merged**